core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala

```diff
@@ -357,30 +380,34 @@ private[spark] class CoarseMesosSchedulerBackend(
    * slave IDs that we might have asked to be killed. It also notifies the driver
    * that an executor was removed.
    */
-  private def executorTerminated(d: SchedulerDriver, slaveId: String, reason: String): Unit = {
+  private def executorTerminated(
+      d: SchedulerDriver,
+      executorId: String,
+      slaveId: String,
+      reason: String): Unit = {
     stateLock.synchronized {
-      if (slaveIdsWithExecutors.contains(slaveId)) {
-        val slaveIdToTaskId = taskIdToSlaveId.inverse()
-        if (slaveIdToTaskId.containsKey(slaveId)) {
-          val taskId: Int = slaveIdToTaskId.get(slaveId)
-          taskIdToSlaveId.remove(taskId)
-          removeExecutor(sparkExecutorId(slaveId, taskId.toString), reason)
+      if (slaveIdsWithExecutors.contains(slaveId) && taskIdToSlaveId.contains(executorId)) {
+        taskIdToSlaveId.remove(executorId)
+        removeExecutor(executorId, reason)
+        val newCount = slaveIdsWithExecutors(slaveId) - 1
+        if (newCount == 0) {
+          slaveIdsWithExecutors.remove(slaveId)
+        } else {
+          slaveIdsWithExecutors(slaveId) = newCount
         }
-        // TODO: This assumes one Spark executor per Mesos slave,
-        // which may no longer be true after SPARK-5095
-        pendingRemovedSlaveIds -= slaveId
-        slaveIdsWithExecutors -= slaveId
       }
     }
   }
 
-  private def sparkExecutorId(slaveId: String, taskId: String): String = {
-    s"$slaveId/$taskId"
-  }
-
   override def slaveLost(d: SchedulerDriver, slaveId: SlaveID): Unit = {
     logInfo(s"Mesos slave lost: ${slaveId.getValue}")
-    executorTerminated(d, slaveId.getValue, "Mesos slave lost: " + slaveId.getValue)
+    // Terminate all executors in the slave
+    stateLock.synchronized {
+      val lostExecutors = taskIdToSlaveId.filter(_._2.equals(slaveId.getValue)).map(_._1)
+      lostExecutors.foreach { taskId =>
+        executorTerminated(d, taskId, slaveId.getValue, "Mesos slave lost: " + slaveId.getValue)
+      }
+    }
   }
 
   override def executorLost(d: SchedulerDriver, e: ExecutorID, s: SlaveID, status: Int): Unit = {
```

Comment:
Shouldn't `executorLost` also get an update? Right now it calls `slaveLost`, which removes all executors on the given slave, but if a slave now runs multiple executors, this seems wrong now.